### PR TITLE
Refunded Products: summarize algorithm

### PR DESCRIPTION
**Pull request by mindgraffiti**
Part 1 of #1717.

So the first step is to get it working. I can successfully smush the data together to create aggregate product cells. It's now readable enough to start making changes 😀 

cc: @koke @jleandroperez (if can spare the time)

### The problem in pictures

The data provided by the API makes the Refunded Products table render like this:

The aggregate data, which is what the user wants to see, should render like this:

### Changes
- I added a new class, `OrderItemRefundSummary`. It's based off of the struct `OrderItemRefund`.
- The `OrderItemRefundSummaryViewModel` is used for configuring the `ProductDetailsTableViewCell`.
- [summedItems](https://github.com/woocommerce/woocommerce-ios/pull/1720/files#diff-ef4a45e82a4b0b7d3eae5834c6f6d07cR18) is where the calculations happen. 

### To Test

The goal is to create multiple refunds with several of the same product type(s) in the refunds.

1. On your test site, have an order with multiple product types. Make sure to note the Order Number.
2. Find the order and "refund by item".
3. Make another refund. Refund a couple more of the same product, and add another product to make it interesting.
4. Make another refund. Refund a couple more of the same product(s) or refund a different item by itself.
5. Check out the develop branch and navigate in the app to: Orders tab > Order Number (Order Details screen). Scroll down a bit until you see the "N Items" cell. [It looks like this](https://user-images.githubusercontent.com/1062444/72114781-45aee200-330a-11ea-8774-cd437ae49d6a.png). Tap the cell to navigate to the Refunded Products screen. Notice that there are multiple cells.
6. Check out this branch. Navigate to the same location. The cells should now be summarized.

### Concerns
I'm still learning about Big O notation. If I understand it right, what I have is approaching (has arrived at?) a quadratic time complexity. Considering I'll have to re-use this algorithm in the Order Details page, that's not good.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift [hunk 1]
```diff
@@ -9,14 +9,88 @@ final class RefundedProductsViewModel {
     ///
     private(set) var order: Order
 
-    /// Array of all refunded items.
+    /// Array of all refunded items from every refund.
     ///
     private(set) var items: [OrderItemRefund]
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: This is where the aggregated data work happens 👇

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift [hunk 2]
```diff
@@ -9,14 +9,88 @@ final class RefundedProductsViewModel {
...
+
+        for productID in uniqueProductIDs {
+            var productGroup = [OrderItemRefund]()
+            var variationIDs = [Int64]()
+
+            // Get every item that has the same productID
+            for item in sorted {
+                if item.productID == productID {
+                    productGroup.append(item)
+                    variationIDs.append(item.variationID)
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: is `variationIDs` used somewhere after this?

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: Thanks for the comments along with the code, it was helpful. If I understand the goals of the refund summary correctly, this is my approach with O(n) with n being the number of refund items:

As I remember, a variation is like a Product and has a unique Product ID, with the logic like in
https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L197 :
```
let productID = item.variationID == 0 ? item.productID : item.variationID
```

(Maybe this can be an extension of `OrderItemRefund` like `item.productOrVariationID`)

so we could probably use this as the identifier for each refund item for a Product or Product Variation.

then we can keep a dictionary mapping from the above `productOrVariationID` to a set or array of refund items:

```
let itemsByProductID = [Int64: [OrderItemRefund]]()
```

then we can populate `itemsByProductID` by iterating over `items` and then add each `item` to the corresponding value given the `productID` calculation.

after `itemsByProductID` is populated, then we can process each Product/Variation ID and summarize all the items into an `OrderItemRefundSummary`.

feel free to correct anything if I misunderstood any part!

Comment (koke) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: I found this part a bit hard to follow, since it has an imperative style and one needs to keep track of what goes into each temporary array.

The way I understand the problem is that you have multiple refund items that you want to group, then show a summary/aggregate of each group. This is a common pattern that I'm used to see as SQL's `GROUP BY` and its aggregate functions (`COUNT`, `SUM`,...), so my background with that might make this proposal biased, but I feel like this would be clearer.

- First, group by the `productID`, `variationID` tuple, and you have one group for each row that you want to present
- Now each group is an array of items, so you need to aggregate their values somehow
- For the item details, take the first item (they should all be identical, right?), then sum the quantities and prices
- Generate an `OrderItemRefundSummary` with that data for each group.

```swift
/// Condense order items into summarized data
///
var summedItems: [OrderItemRefundSummary] {
    /// OrderItemRefund.orderItemID isn't useful for finding duplicates here,
    /// because multiple refunds cause orderItemIDs to be unique.
    /// Instead, we need to find duplicate *Products*.
    let currency = CurrencyFormatter()

    let grouped = Dictionary(grouping: items) { (item) in
        // This needs to be hashable so a tuple won't do it
        // I doubt this would become a performance issue, but it could be replaced
        // by a hashable struct
        return "\(item.productID)-\(item.variationID)"
    }
    return grouped.compactMap { (_, items) in
        // Here we iterate over each group's items

        // All items should be equal except for quantity and price, so we pick the first
        guard let item = items.first else {
            // This should never happen, but let's be safe
            return nil
        }

        // Sum the quantities
        let totalQuantity = items
            .map( { $0.quantity } )
            .reduce(0, +)
        // Sum the refunded amount
        let totalTax = items
            .compactMap( { currency.convertToDecimal(from: $0.totalTax) } )
            .reduce(NSDecimalNumber(value: 0), { $0.adding($1) })

        return OrderItemRefundSummary(
            name: item.name,
            productID: item.productID,
            variationID: item.variationID,
            quantity: totalQuantity,
            price: item.price,
            sku: item.sku,
            totalTax: totalTax)
    }
}
```

I feel like key paths could make the syntax a bit better, if we started adding a few helper extensions:

```swift
extension Sequence {
    func map<T>(_ keyPath: KeyPath<Element, T>) -> [T] {
        return map { $0[keyPath: keyPath] }
    }

    func sum<T: Numeric>(_ keyPath: KeyPath<Element, T>) -> T {
        return map(keyPath).sum()
    }
}

extension Sequence where Element: Numeric {
    /// Returns the sum of all elements in the collection
    func sum() -> Element { return reduce(0, +) }
}
```

Then the total quantity could be calculated as:

```swift
let totalQuantity = items.sum(\.quantity)
```

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift [hunk 3]
```diff
@@ -43,6 +43,7 @@ final class OrderDetailsViewModel {
     var refundedItems: [OrderItemRefund] {
         let refunds = dataSource.refunds
         var items = [OrderItemRefund]()
+
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift]: Yeah, I took out a bunch of experimental code before creating the PR. I was improving the readability while re-reading this class. I wish Hound could catch more of our code style rules. I would have added the return sooner.

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummaryViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
+import Foundation
+import Yosemite
+
+
+// MARK: - OrderItemRefundSummary View Model
+//
+struct OrderItemRefundSummaryViewModel {
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummaryViewModel.swift]: Thanks! Switched to default file access for now. It was still declared as public for experiments that I've since deleted 😅

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummaryViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                                              comment: "<item total> (<item individual price> multipled by <quantity>)")
+        let priceText = String.localizedStringWithFormat(priceTemplate, itemTotal, itemPrice, quantity)
+
+        return priceText
+    }
+
+    /// Item's Tax
+    /// Return $0.00 if there is no tax.
+    ///
+    var tax: String {
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummaryViewModel.swift]: Good catch! It definitely should not be part of this PR, because I'm not sure what the Order Details > Product / Qty changes will be.

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift (hunk 1 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: Some quick notes, as I work my way through this 😀

>a variation is like a Product and has a unique Product ID

I think that is true for Products (a product that was pulled from the products results controller), but not true for Order Items (as I encountered that issue early when writing this PR). `OrderItem`s have unique productIDs, but the variationID is only unique within the context of a productID. So for example,  an `OrderItem` that was derived from a variable product `Blue Fish - SM` can have the same variationID (productID: 71, variationID: 22) as `Pink Penguin - L` (productID: 3, variationID: 22). 

> I found this part a bit hard to follow, since it has an imperative style and one needs to keep track of what goes into each temporary array.

Yes. I fell back on Objective-C thinking to create a working example. This is very ... _not_ Swift

Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: @jaclync thank you for proposing a better solution than what I had 🌟. I really appreciate it. @koke thank you for writing out an example of what you talked about before. I couldn't make the leap from what you explained to actually putting it into code.

I decided to go with @koke's solution. While researching how `Dictionary(grouping:by)` works, I learned that returning a hashable value creates an O(1) result (doesn't matter how many order items we're grouping)!

The code below is ready for another review.

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsDataSource.swift [hunk 6]
```diff
@@ -146,6 +146,15 @@ extension RefundedProductsDataSource {
 
         sections = [refundedProducts].compactMap { $0 }
     }
+
+    /// Handle taps on cells
+    ///
+    func tableView(_ tableView: UITableView,
+                   in viewController: UIViewController,
+                   didSelectRowAt indexPath: IndexPath) {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Refunds/RefundedProductsDataSource.swift]: are the diffs here trying to disable table view cell tap actions? if so, the cells are still tappable from testing that navigate to a Product modal. It looks like `RefundedProductsViewController` is the class that implements `UITableViewDelegate` that handles the tap action: https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Payment%20Section/Refunds/RefundedProductsViewController.swift#L152-L156

alternatively, you can set [`UITableView`'s `allowsSelection` property](https://developer.apple.com/documentation/uikit/uitableview/1614911-allowsselection) to false to disable cell tap action for all the rows

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: @koke's solution looks great and elegant 👍 TIL about the key paths mapping and `Dictionary(grouping:by:)`!

> a variation is like a Product and has a unique Product ID
> So for example, an OrderItem that was derived from a variable product Blue Fish - SM can have the same variationID (productID: 71, variationID: 22) as Pink Penguin - L (productID: 3, variationID: 22).

from my understanding, variation IDs and product IDs are unique per site, because they are all `Post`s behind the scene. Local product attributes are unique per product.

File: WooCommerce/Classes/Extensions/Array+Helpers.swift [hunk 7]
```diff
@@ -17,11 +16,33 @@ extension Array {
...
+extension Sequence {
+    /// Get the keypaths for a elemtents in a sequence.
+    ///
+    func map<T>(_ keyPath: KeyPath<Element, T>) -> [T] {
+        return map { $0[keyPath: keyPath] }
+    }
+
+    /// Sum a sequence of elements by keypath.
+    func sum<T: Numeric>(_ keyPath: KeyPath<Element, T>) -> T {
+        return map(keyPath).sum()
```
Comment (jaclync) [inline WooCommerce/Classes/Extensions/Array+Helpers.swift]: these key paths mapping and reducing are really cool!

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+//
+extension OrderItemRefundSummary: Comparable {
+    public static func == (lhs: OrderItemRefundSummary, rhs: OrderItemRefundSummary) -> Bool {
+        return lhs.productID == rhs.productID &&
+            lhs.variationID == rhs.variationID
+    }
+
+    public static func < (lhs: OrderItemRefundSummary, rhs: OrderItemRefundSummary) -> Bool {
+        return lhs.productID < rhs.productID ||
+            (lhs.productID == rhs.productID && lhs.name < rhs.name)
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift]: since the `==` uses both `productID` and `variationID`, does it make sense for the second condition to also check `variationID`?

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+}
+
+
+// MARK: - Hashable Conformance
+//
+extension OrderItemRefundSummary: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(productID)
+        hasher.combine(variationID)
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift]: is `OrderItemRefundSummary` hashed in this PR? if not, maybe it's not necessary to implement `Hashable`?

File: WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift (hunk 1 above)
Comment (jleandroperez) [reply WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift]: @koke TIL on the `Dictionary(groping:by:)` as well, that one is VERY cool!!

@mindgraffiti apologies I'm late to the party!!! I would probably have gone in the lines of splitting the `[OrderItemRefund]` property into multiple smaller pieces (and maybe unit test that!).

**BUT** I think Koke's SQL-like approach is as elegant and readable as it gets. I think he really nailed it there!

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+final class OrderItemRefundSummary {
+    let productID: Int64
+    let variationID: Int64
+
+    let name: String
+
+    /// Price, total, and totalTax are currency values.
+    /// When handling currencies, `NSDecimalNumber` is a powerhouse
+    /// for localization and string-to-number conversions.
+    /// `Decimal` doesn't have all of the `NSDecimalNumber` APIs (yet).
```
Comment (koke) [inline WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift]: TIL. I was going to suggest `Decimal` because it conforms to `Numeric` and you can use `+` and `sum` as well for more clarity. Which APIs are missing? Could you use `Decimal` by default and cast `as NSDecimalNumber` when you need the extra APIs?

Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift]: Good question! The short answer is: I didn't know any better at the time and I can't use Decimal here without refactoring the `CurrencyFormatter` class first.

Originally I chose `NSDecimalNumber` because:
1) The models are backed by a core data store, and the `NSManagedObject`s were `NSDecimalNumber` types. (Later I learned that I could have defined the `NSManagedObject`s to be Swift types instead.) 
2) While writing the CurrencyFormatter I came across limitations to `Decimal` (which sadly, I didn't leave my future-self enough hints about) and continued to use the `NSDecimalNumber` type in it. This could probably be re-visited. 

The [CurrencyFormatter class](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/Tools/Currency/CurrencyFormatter.swift) would need refactored to use `Decimal` before I could use `Decimal` here, because all currencies that were strings from the API have to use the [convertToDecimal](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/Tools/Currency/CurrencyFormatter.swift#L12-L27) method first. Explainer for that method, ref: p7bje6-1Ml-p2.

File: WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift (hunk 8 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Refunds/OrderItemRefundSummary.swift]: Oops. Yep that's correct 🤦‍♀  

The `name` shouldn't be compared, the `variationID` should be.

**Comment by mindgraffiti** [comment]
✅ Fixed: tapping on a refunded Product goes to the Product details, not sure if this is expected
✅ Fixed: the order of the refunded products seems to be nondeterministic, might be nice to sort them
✅ questions answered
✅ all other code changes complete

TODO: add unit tests in a future PR

Merging this one now.
